Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2024 Bjørn Kjos-Hanssen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bjørn Kjos-Hanssen
+-/
+import Mathlib.Analysis.Calculus.MeanValue
+import Mathlib.Order.Interval.Set.Basic
+import Mathlib.Topology.Defs.Filter
+import Mathlib.Topology.Order.OrderClosedExtr
+import Mathlib.Analysis.Calculus.FDeriv.Add
+/-!
+# The First-Derivative Test
+
+We prove the first-derivative test in the strong form given on Wikipedia.
+
+The test is proved over the real numbers ℝ
+using `monotoneOn_of_deriv_nonneg` from [Mathlib.Analysis.Calculus.MeanValue].
+
+## Main results
+
+* `first_derivative_test_max`: Suppose `f` is a real-valued function of a real variable
+  defined on some interval containing the point `a`.
+  Further suppose that `f` is continuous at `a` and differentiable on some open interval
+  containing `a`, except possibly at `a` itself.
+
+  If there exists a positive number `r > 0` such that for every `x` in `(a − r, a)`
+  we have `f′(x) ≥ 0`, and for every `x` in `(a, a + r)` we have `f′(x) ≤ 0`,
+  then `f` has a local maximum at `a`.
+
+* `first_derivative_test_min`: The dual of `first_derivative_max`, for minima.
+
+* `first_derivative_test_max'`: A version of `first_derivative_test_max` for filters.
+
+## Tags
+
+derivative test, calculus
+-/
+
+open Set Topology
+
+/-!
+### Some facts about differentiability and continuity
+
+We prove a couple of auxiliary lemmas elaborating on facts such as
+"differentiable implies continuous",
+"an open interval is an open set", and "`fun x => -x` is antitone". -/
+
+/-- If `f'` is the derivative of `f` then  `f' x ≤ 0 → 0 ≤ (-f)' x`. -/
+theorem deriv_neg_nonneg {f : ℝ → ℝ} {a b : ℝ} (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0) {x : ℝ} (hx : x ∈ Set.Ioo a b) : 0 ≤ deriv (-f) x :=
+  (@deriv.comp ℝ _ x ℝ _ _ f (fun x => -x)
+    (Differentiable.differentiableAt differentiable_neg)
+    (DifferentiableOn.differentiableAt hd₀ (Ioo_mem_nhds hx.1 hx.2))) ▸ (by
+    rw [deriv_neg'', neg_mul, one_mul, Left.nonneg_neg_iff];
+    exact h₀ _ hx
+  )
+
+/-- If `f'` is the derivative of `f` then  `0 ≤ f' x → (-f)' x ≤ 0`. -/
+theorem deriv_neg_nonpos {f : ℝ → ℝ} {b c : ℝ} (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) {x : ℝ} (hx : x ∈ Set.Ioo b c) : deriv (-f) x ≤ 0 :=
+    (@deriv.comp ℝ _ x ℝ _ _ f (fun x => -x)
```

Comment:
You can use named arguments to avoid Morse code.

Reply:
Thanks, that's useful.